File: qbopt/analysis/liveness.py
"""Which values are live where, over MIR's own SSA values.

Nothing here names a register. It sat in regalloc.py because that is what
first needed it, and a MIR pass asking regalloc a liveness question is a
MIR pass calling a machine one -- rule 5's own example. The allocator is
one caller of this; the hoist and the fold are others.
"""

from dataclasses import dataclass

from qbopt.model import mir
from qbopt.model.mir import Value


@dataclass(frozen=True, slots=True)
class Liveness:
    live_in: dict[int, frozenset[Value]]
    live_out: dict[int, frozenset[Value]]


def _defines(block: mir.MirBlock) -> set[Value]:
    return {phi.result for phi in block.phis} | {one for op in block.ops for one in op.defines}


def _exposed(block: mir.MirBlock) -> set[Value]:
    """Values the ordinary operations read before writing.

    A phi's arguments belong to the edges they arrive on, so they are the
    predecessor's business and never this block's. Phi results remain here
    when an operation reads them: that demand decides whether the matching
    edge argument is live at all.
    """
    live: set[Value] = set()
    for op in reversed(block.ops):
        live -= set(op.defines)
        live |= set(op.uses)
    return live


def entry_values(body: mir.MirBody) -> frozenset[Value]:
    """Values the caller supplied: used somewhere, defined by nothing here.

    A body reads a register before writing it whenever BC passes something
    in, and mir._Namer invents a value for it rather than pretending the
    read has no operand. Nothing in the body defines one, so unless the
    entry block is told it does, backward liveness never kills it: it stays
    live at every point that can reach its use, which around a loop is
    everywhere. That reported twelve values live at once in a body BC
    compiled into six registers -- impossible, and the reason pressure()
    treats a number above len(AVAILABLE) as a bug rather than a spill.
    """
    defined = {one for block in body.blocks for one in _defines(block)}
    used = {
        one
        for block in body.blocks
        for one in (
            {u for op in block.ops for u in (*op.uses, *op.exits)}
            | {v for phi in block.phis for v in phi.incoming.values()}
        )
    }
    return frozenset(used - defined)


def pressure(
    body: mir.MirBody, found: "Liveness | None" = None, inside: "frozenset[int] | set[int] | None" = None
) -> int:
    """The most values live at once, flags aside -- anywhere, or in `inside`.

    Cannot exceed the number of registers BC itself used, because the
    program came out of them, so a number above the register file is a bug
    in the liveness and not a body that needs spilling.

    `inside` asks it of one loop's blocks, which is what a pass deciding
    whether it can afford another loop-carried value has to know: a body's
    peak says nothing about the loop the value would live around.
    """
    found = found or live(body)
    peak = 0
    for block in body.blocks:
        if inside is not None and block.at not in inside:
            continue
        alive = set(found.live_out[block.at])
        peak = max(peak, len([one for one in alive if not one.flags]))
        for op in reversed(block.ops):
            alive -= set(op.defines)
            alive |= set(op.uses)
            peak = max(peak, len([one for one in alive if not one.flags]))
    return peak


def phi_inputs(body: mir.MirBody, found: "Liveness | None" = None) -> frozenset[Value]:
    """The edge operands of phis whose results are actually live.

    SSA construction may create phis for machine state which no operation
    observes.  Such a phi is not a use of every incoming value.  Walk from
    ordinary live-out back to the point immediately after the phis, then
    select only the incoming edges of demanded results.
    """
    found = found or live(body)
    inputs: set[Value] = set()
    for block in body.blocks:
        alive = set(found.live_out[block.at])
        for op in reversed(block.ops):
            alive.difference_update(op.defines)
            alive.update(op.uses)
        inputs.update(incoming for phi in block.phis if phi.result in alive for incoming in phi.incoming.values())
    return frozenset(inputs)


def live(body: mir.MirBody) -> Liveness:
    """What is live at each block's entry and exit, to a fixed point."""
    # Phi definitions happen before ordinary operations and read one selected
    # predecessor edge. Keeping the two kinds of definition separate avoids
    # making every syntactic phi input live when the phi result is dead.
    op_defines = {block.at: {one for op in block.ops for one in op.defines} for block in body.blocks}
    phi_defines = {block.at: {phi.result for phi in block.phis} for block in body.blocks}
    exposed = {block.at: _exposed(block) for block in body.blocks}
    if body.blocks:
        # Defined at the top of the entry block, before its first
        # instruction: added to what it defines AND removed from what it
        # leaves upward-exposed. Only the first is not enough -- live_in is
        # (live_out - defines) + exposed, so an exposed use puts the value
        # straight back and the kill never happens.
        arriving = set(entry_values(body))
        op_defines[body.entry] |= arriving
        exposed[body.entry] -= arriving
    live_in: dict[int, set[Value]] = {block.at: set() for block in body.blocks}
    live_out: dict[int, set[Value]] = {block.at: set() for block in body.blocks}
    after_phis: dict[int, set[Value]] = {block.at: set() for block in body.blocks}

    changing = True
    while changing:
        changing = False
        for block in body.blocks:
            # An exit observation happens after the operation carrying it.
            # Seeding live-out, rather than treating it as an operand, lets a
            # terminal call define the value its caller observes without also
            # pretending that the call reads its own result.
            out: set[Value] = {value for op in block.ops for value in mir.exit_values(op)}
            for successor in block.succ:
                found = body.block(successor)
                if found is None:
                    continue
                out |= live_in[successor]
                out |= {
                    phi.incoming[block.at]
                    for phi in found.phis
                    if phi.result in after_phis[successor] and block.at in phi.incoming
                }
            after = (out - op_defines[block.at]) | exposed[block.at]
            inside = after - phi_defines[block.at]
            if out != live_out[block.at] or inside != live_in[block.at] or after != after_phis[block.at]:
                live_out[block.at], live_in[block.at], after_phis[block.at] = out, inside, after
                changing = True

    return Liveness(
        {at: frozenset(what) for at, what in live_in.items()},
        {at: frozenset(what) for at, what in live_out.items()},
    )

File: src/analysis/liveness.rs
//! Port of `qbopt/analysis/liveness.py`: which values are live where, over
//! MIR's own SSA values.

use std::collections::{BTreeMap, BTreeSet};

use crate::model::mir::{self, MirBlock, MirBody, Value};

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Liveness {
    pub live_in: BTreeMap<i64, BTreeSet<Value>>,
    pub live_out: BTreeMap<i64, BTreeSet<Value>>,
}

fn _defines(block: &MirBlock) -> BTreeSet<Value> {
    let mut out: BTreeSet<Value> = block.phis.iter().map(|phi| phi.result).collect();
    out.extend(block.ops.iter().flat_map(|op| op.defines.iter().copied()));
    out
}

/// Values the ordinary operations read before writing.
fn _exposed(block: &MirBlock) -> BTreeSet<Value> {
    let mut live = BTreeSet::new();
    for op in block.ops.iter().rev() {
        for one in &op.defines {
            live.remove(one);
        }
        live.extend(op.uses.iter().copied());
    }
    live
}

/// Values the caller supplied: used somewhere, defined by nothing here.
pub fn entry_values(body: &MirBody) -> BTreeSet<Value> {
    let defined: BTreeSet<Value> = body.blocks.iter().flat_map(_defines).collect();
    let mut used = BTreeSet::new();
    for block in &body.blocks {
        for op in &block.ops {
            used.extend(op.uses.iter().copied());
            used.extend(op.exits.iter().copied());
        }
        for phi in &block.phis {
            used.extend(phi.incoming.values().copied());
        }
    }
    used.difference(&defined).copied().collect()
}

/// The most values live at once, flags aside -- anywhere, or in `inside`.
pub fn pressure(body: &MirBody, found: Option<&Liveness>, inside: Option<&BTreeSet<i64>>) -> usize {
    let owned;
    let found = match found {
        Some(found) => found,
        None => {
            owned = live(body);
            &owned
        }
    };
    let count = |alive: &BTreeSet<Value>| alive.iter().filter(|one| !one.flags).count();
    let mut peak = 0;
    for block in &body.blocks {
        if inside.is_some_and(|inside| !inside.contains(&block.at)) {
            continue;
        }
        let mut alive = found.live_out[&block.at].clone();
        peak = peak.max(count(&alive));
        for op in block.ops.iter().rev() {
            for one in &op.defines {
                alive.remove(one);
            }
            alive.extend(op.uses.iter().copied());
            peak = peak.max(count(&alive));
        }
    }
    peak
}

/// The edge operands of phis whose results are actually live.
pub fn phi_inputs(body: &MirBody, found: Option<&Liveness>) -> BTreeSet<Value> {
    let owned;
    let found = match found {
        Some(found) => found,
        None => {
            owned = live(body);
            &owned
        }
    };
    let mut inputs = BTreeSet::new();
    for block in &body.blocks {
        let mut alive = found.live_out[&block.at].clone();
        for op in block.ops.iter().rev() {
            for one in &op.defines {
                alive.remove(one);
            }
            alive.extend(op.uses.iter().copied());
        }
        for phi in &block.phis {
            if alive.contains(&phi.result) {
                inputs.extend(phi.incoming.values().copied());
            }
        }
    }
    inputs
}

/// What is live at each block's entry and exit, to a fixed point.
pub fn live(body: &MirBody) -> Liveness {
    let mut op_defines: BTreeMap<i64, BTreeSet<Value>> = body
        .blocks
        .iter()
        .map(|block| (block.at, block.ops.iter().flat_map(|op| op.defines.iter().copied()).collect()))
        .collect();
    let phi_defines: BTreeMap<i64, BTreeSet<Value>> =
        body.blocks.iter().map(|block| (block.at, block.phis.iter().map(|phi| phi.result).collect())).collect();
    let mut exposed: BTreeMap<i64, BTreeSet<Value>> =
        body.blocks.iter().map(|block| (block.at, _exposed(block))).collect();
    if !body.blocks.is_empty() {
        let arriving = entry_values(body);
        op_defines.entry(body.entry).or_default().extend(arriving.iter().copied());
        if let Some(exposed) = exposed.get_mut(&body.entry) {
            exposed.retain(|one| !arriving.contains(one));
        }
    }
    let empty = || body.blocks.iter().map(|block| (block.at, BTreeSet::new())).collect::<BTreeMap<_, _>>();
    let mut live_in = empty();
    let mut live_out = empty();
    let mut after_phis = empty();

    let mut changing = true;
    while changing {
        changing = false;
        for block in &body.blocks {
            let mut out: BTreeSet<Value> =
                block.ops.iter().flat_map(|op| mir::exit_values(op).iter().copied()).collect();
            for successor in &block.succ {
                let Some(found) = body.block(*successor) else {
                    continue;
                };
                out.extend(live_in[successor].iter().copied());
                for phi in &found.phis {
                    if after_phis[successor].contains(&phi.result) {
                        out.extend(phi.incoming.get(&block.at).copied());
                    }
                }
            }
            let mut after: BTreeSet<Value> = out.difference(&op_defines[&block.at]).copied().collect();
            after.extend(exposed[&block.at].iter().copied());
            let inside: BTreeSet<Value> = after.difference(&phi_defines[&block.at]).copied().collect();
            if out != live_out[&block.at] || inside != live_in[&block.at] || after != after_phis[&block.at] {
                live_out.insert(block.at, out);
                live_in.insert(block.at, inside);
                after_phis.insert(block.at, after);
                changing = true;
            }
        }
    }
    Liveness { live_in, live_out }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::model::mir::Phi;

    /// PARITYCONTROL's dead join-flag phis kept ADD/ADC live as word operations.
    #[test]
    fn test_a_dead_phi_does_not_keep_its_edge_operand_live() {
        let incoming = Value { flags: true, ..Value::new(1, 0x10) };
        let merged = Value { flags: true, ..Value::new(2, 0x20) };
        let phi = Phi { result: merged, incoming: [(0x10, incoming)].into_iter().collect() };
        let body = MirBody::new(
            0x10,
            vec![MirBlock::new(0x10, vec![], vec![], vec![0x20]), MirBlock::new(0x20, vec![phi], vec![], vec![])],
        );
        let found = live(&body);
        assert!(!found.live_out[&0x10].contains(&incoming));
    }
}
